files/private-chef-ctl-commands/key-control.rb

```diff
@@ -0,0 +1,255 @@
+
+  id = nil
+  DB.fetch("SELECT id FROM clients WHERE name='#{cmd_args[1]}'") do |row|
+    id = row[:id]
+  end
+  unless id
+    STDERR.puts "Error: client could not be found for clientname #{cmd_args[1]}"
+    exit 1
+  end
+
+  org_id = nil
+  DB.fetch("SELECT id FROM orgs WHERE name='#{cmd_args[0]}'") do |row|
+    org_id = row[:id]
+  end
+  unless org_id
+    STDERR.puts "Error: organization could not be found for orgname #{cmd_args[0]}"
+    exit 1
+  end
+
+  # delete orgname from the cmd_args so that it has the same interface as add-user-key
+  cmd_args.delete_at(0)
+
+  addkey.add_key(cmd_args, id)
+end
+
+add_command_under_category "add-user-key", "key-rotation", "Create a new user key", 2 do
+  addkey = AddKey.new
+  addkey.check_org_args(cmd_args, "Usage: chef-server-ctl add-user-key <username> <path_to_public_key> [--expiration-date <date>] [--key-name <string>]")
+  id = nil
+  DB.fetch("SELECT id FROM users WHERE username='#{cmd_args[0]}'") do |row|
+    id = row[:id]
+  end
+  unless id
+    STDERR.puts "Error: user could not be found for username #{cmd_args[0]}"
+    exit 1
+  end
+  AddKey.new.add_key(cmd_args, id)
+end
+
+
+add_command_under_category "delete-key", "key-rotation", "Delete a key", 2 do
+    unless cmd_args.length == 2 || \
+      (cmd_args.length == 4 && (cmd_args[2] == "--expiration-date" || cmd_args[2] == "--key-name")) || \
+      (cmd_args.length == 6 && ((cmd_args[2] == "--expiration-date" && cmd_args[4] == "--key-name") || (cmd_args[4] == "--expiration-date" && cmd_args[2] == "--key-name")))
+      STDERR.puts usage_msg
+      STDERR.puts "Usage: Expiration date defaults to infinite. Pass an ISO 8601 fomatted string: YYYY-MM-DDTHH:MM:SS e.g. 2013-12-24T21:00:00"
+      STDERR.puts "Usage: Default name used is the fingerprint of the key passed."
+      exit 1
+    end
+
+  unless cmd_args.length == 2
+    STDERR.puts "Usage: chef-server-ctl delete-key <id> <key_id>"
+    exit 1
+  end
+
+  results = DB[:keys].where("id='#{cmd_args[0]}' AND key_id='#{cmd_args[1]}'")
+  if results.count == 0
+    STDERR.puts "Error: Could not find key where id #{cmd_args[0]} and key_id #{cmd_args[1]}"
+    exit 1
+  end
+
+  # this is just a safety check so that we don't accidentally delete multiple rows, should never happen
+  if results.count > 1
+    STDERR.puts "Error: Multiple keys found for id #{cmd_args[0]} and key_id #{cmd_args[1]}. Please contact a sysadmin."
+    exit 1
+  end
+
+  results.delete
+end
+
+add_command_under_category "list-client-keys", "key-rotation", "List keys for a user", 2 do
+  # allow users to hide public key from output, this will allow for much easier
+  # viewing of keys if a user has lots of keys and they just want to see the key_ids
+  hide_public_keys = cmd_args[2] == "--hide-public-keys"
+  unless cmd_args.length == 2 || (cmd_args.length == 3 && cmd_args[2] == "--hide-public-keys")
+    STDERR.puts "Usage: chef-server-ctl user-key-list <organization> <clientname> [--hide-public-keys]"
```

Comment:
This is why copy-pasta is bad. Again, I thought about pull those all into variables after the fact, but this code is going away for the most part. Although this specifically is probably an example of part of the code that isn't.